extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php

```diff
@@ -0,0 +1,254 @@
+
+	const JWT_AUTH_TOKEN_COOKIE_NAME = 'jp-premium-content-session';
+	const DECODE_EXCEPTION_FEATURE   = 'memberships';
+	const DECODE_EXCEPTION_MESSAGE   = 'Problem decoding provided token';
+	const REST_URL_ORIGIN            = 'https://subscribe.wordpress.com/';
+
+	/**
+	 * Initialize the token subscription service.
+	 *
+	 * @inheritDoc
+	 */
+	public function initialize() {
+		$token = $this->token_from_request();
+		if ( null !== $token ) {
+			$this->set_token_cookie( $token );
+		}
+	}
+
+	/**
+	 * The user is visiting with a subscriber token cookie.
+	 *
+	 * This is theoretically where the cookie JWT signature verification
+	 * thing will happen.
+	 *
+	 * How to obtain one of these (or what exactly it is) is
+	 * still a WIP (see api/auth branch)
+	 *
+	 * @inheritDoc
+	 *
+	 * @param array $valid_plan_ids List of valid plan IDs.
+	 */
+	public function visitor_can_view_content( $valid_plan_ids ) {
+
+		// URL token always has a precedence, so it can overwrite the cookie when new data available.
+		$token = $this->token_from_request();
+		if ( $token ) {
+			$this->set_token_cookie( $token );
+		} else {
+			$token = $this->token_from_cookie();
+		}
+
+		$is_valid_token = true;
+
+		if ( empty( $token ) ) {
+			// no token, no access.
+			$is_valid_token = false;
+		} else {
+			$payload = $this->decode_token( $token );
+			if ( empty( $payload ) ) {
+				$is_valid_token = false;
+			}
+		}
+
+		if ( $is_valid_token ) {
+			$subscriptions = (array) $payload['subscriptions'];
+		} elseif ( is_user_logged_in() ) {
+			// If there is no token, but the user is logged in, get current subscriptions and determine if the user has
+			// a valid subscription to match the plan ID.
+			$subscriptions = apply_filters( 'earn_get_user_subscriptions_for_site_id', array(), wp_get_current_user()->ID, $this->get_site_id() );
+			if ( empty( $subscriptions ) ) {
+				return false;
+			}
+			// format the subscriptions so that they can be validated.
+			$subscriptions = self::abbreviate_subscriptions( $subscriptions );
+		} else {
+			return false;
+		}
+
+		return $this->validate_subscriptions( $valid_plan_ids, $subscriptions );
+	}
+
+	/**
+	 * Decode the given token.
+	 *
+	 * @param string $token Token to decode.
+	 *
+	 * @return array|false
+	 */
+	public function decode_token( $token ) {
+		try {
+			$key = $this->get_key();
+			return $key ? (array) JWT::decode( $token, $key, array( 'HS256' ) ) : false;
+		} catch ( \Exception $exception ) {
+			// TODO: The SignatureInvalidException is in a different namespace from JWT, so if we want to catch that
+			// one specifically then we'll have to alias it.
+			$logstash = array(
+				'feature' => self::DECODE_EXCEPTION_FEATURE,
+				'message' => self::DECODE_EXCEPTION_MESSAGE,
+				'extra'   => wp_json_encode( compact( 'exception', 'token' ) ),
+			);
+         // phpcs:ignore ImportDetection.Imports.RequireImports.Symbol
+			log2logstash( $logstash );
```

Comment:
Could this be an issue on non-WordPress.com environments?

Reply:
Removed :+1: